### PR TITLE
feat(manager): maintain n running nodes

### PR DESCRIPTION
**Pull request by RolandSherwin**
[feat(manager): maintain n running nodes](https://github.com/maidsafe/safe_network/commit/994ddbc32c945e8d1a0a0168b45b09a1dac1a3ae)
- This commit was supposed to be part of #1694  as mentioned [here](https://github.com/maidsafe/safe_network/pull/1694#pullrequestreview-2049047244) . But it got removed during a rebase. So adding it back in.

[feat(launchpad): provide safenode path for testing](https://github.com/maidsafe/safe_network/commit/a8ce64fd001880d425c614429aa8bd16bb611a98)
- This commit is used to use a pre built safenode binary inside launchpad for testing things out. Else we'd be fetching the nodes from latest release everytime (which will be old).

### CODE REVIEW

File: node-launchpad/src/components/home.rs [hunk 1]
```diff
@@ -469,35 +388,19 @@ fn add_and_start_nodes(
...
+            VerbosityLevel::Minimal,
             NODE_START_INTERVAL as u64,
-            vec![],
-            nodes_to_start,
-            sn_node_manager::VerbosityLevel::Minimal,
         )
         .await
         {
-            error!("Error while starting services {err:?}");
+            error!("Error while maintaining n running nodes {err:?}");
```
Comment (maqi) [inline node-launchpad/src/components/home.rs]: this `n` can be replaced by `count` number ?

File: node-launchpad/src/components/home.rs [hunk 2]
```diff
@@ -469,35 +388,19 @@ fn add_and_start_nodes(
...
         } else {
-            info!("Successfully started services");
+            info!("Maintain node run count successful.");
```
Comment (maqi) [inline node-launchpad/src/components/home.rs]: Maintain {count} running nodes successfully

File: sn_node_manager/src/cmd/node.rs [hunk 3]
```diff
@@ -492,6 +492,138 @@ pub async fn upgrade(
     Ok(())
 }
 
+/// This functions makes sure that we run exactly the provided count nodes at any time by stopping nodes or
```
Comment (maqi) [inline sn_node_manager/src/cmd/node.rs]: `This functions makes sure that we run exactly the provided count nodes at any time` =>
`This function ensures running exactly the provided number of nodes at any time`

File: sn_node_manager/src/cmd/node.rs [hunk 4]
```diff
@@ -492,6 +492,138 @@ pub async fn upgrade(
...
+    rpc_address: Option<Ipv4Addr>,
+    rpc_port: Option<PortRange>,
+    src_path: Option<PathBuf>,
+    url: Option<String>,
+    upnp: bool,
+    user: Option<String>,
+    version: Option<String>,
+    verbosity: VerbosityLevel,
+    start_node_interval: u64,
+) -> Result<()> {
```
Comment (maqi) [inline sn_node_manager/src/cmd/node.rs:520]: I am suprised this doesn't trigger a clippy complain that `too many arguments to be passed` 
maybe make those for `add` as a separate struct ?

**Review by maqi — APPROVED** [review]

**Review by mazzi — APPROVED** [review]
LGTM 🚀

File: sn_node_manager/src/cmd/node.rs (hunk 4 above)
Comment (RolandSherwin) [reply sn_node_manager/src/cmd/node.rs:520]: I think there is a global `allow()` macro. Most of the functions there has got like 10+ args. Might be out of scope for this PR. :/

**Pull request merged**